File: decompiled/sources/com/hjq/widget/view/SubmitButton.java
package com.hjq.widget.view;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathMeasure;
import android.graphics.PorterDuff;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.animation.AccelerateInterpolator;
import androidx.appcompat.widget.AppCompatButton;
import com.hjq.widget.R;

public final class SubmitButton extends AppCompatButton {
    private static final int STATE_LOADING = 2;
    private static final int STATE_NONE = 0;
    private static final int STATE_RESULT = 3;
    private static final int STATE_SUBMIT = 1;
    private static final int STYLE_LOADING = 0;
    private static final int STYLE_PROGRESS = 1;
    private Paint mBackgroundPaint;
    private Path mButtonPath;
    private int mButtonState;
    private RectF mCircleLeft;
    private RectF mCircleMid;
    private RectF mCircleRight;
    private float mCurrentProgress;
    /* access modifiers changed from: private */
    public boolean mDoResult;
    private Path mDstPath;
    private final int mErrorColor;
    private Path mLoadPath;
    private float mLoadValue;
    private ValueAnimator mLoadingAnim;
    private Paint mLoadingPaint;
    private int mMaxHeight;
    private int mMaxWidth;
    private PathMeasure mPathMeasure;
    private final int mProgressColor;
    private final int mProgressStyle;
    private ValueAnimator mResultAnim;
    private Paint mResultPaint;
    private Path mResultPath;
    private ValueAnimator mSubmitAnim;
    private boolean mSucceed;
    private final int mSucceedColor;
    private int mViewHeight;
    private int mViewWidth;
    private int mX;
    private int mY;

    public SubmitButton(Context context) {
        this(context, (AttributeSet) null);
    }

    public SubmitButton(Context context, AttributeSet attributeSet) {
        this(context, attributeSet, 0);
    }

    public SubmitButton(Context context, AttributeSet attributeSet, int i) {
        super(context, attributeSet, i);
        this.mButtonState = 0;
        setLayerType(1, (Paint) null);
        TypedArray obtainStyledAttributes = context.obtainStyledAttributes(attributeSet, R.styleable.SubmitButton, i, 0);
        this.mProgressColor = obtainStyledAttributes.getColor(R.styleable.SubmitButton_progressColor, getAccentColor());
        this.mSucceedColor = obtainStyledAttributes.getColor(R.styleable.SubmitButton_succeedColor, Color.parseColor("#19CC95"));
        this.mErrorColor = obtainStyledAttributes.getColor(R.styleable.SubmitButton_errorColor, Color.parseColor("#FC8E34"));
        this.mProgressStyle = obtainStyledAttributes.getInt(R.styleable.SubmitButton_progressStyle, 0);
        obtainStyledAttributes.recycle();
        initPaint();
        resetPaint();
    }

    private void initPaint() {
        this.mBackgroundPaint = new Paint();
        this.mLoadingPaint = new Paint();
        this.mResultPaint = new Paint();
        this.mButtonPath = new Path();
        this.mLoadPath = new Path();
        this.mResultPath = new Path();
        this.mDstPath = new Path();
        this.mCircleMid = new RectF();
        this.mCircleLeft = new RectF();
        this.mCircleRight = new RectF();
        this.mPathMeasure = new PathMeasure();
    }

    private void resetPaint() {
        this.mBackgroundPaint.setColor(this.mProgressColor);
        this.mBackgroundPaint.setStrokeWidth(5.0f);
        this.mBackgroundPaint.setAntiAlias(true);
        this.mLoadingPaint.setColor(this.mProgressColor);
        this.mLoadingPaint.setStyle(Paint.Style.STROKE);
        this.mLoadingPaint.setStrokeWidth(9.0f);
        this.mLoadingPaint.setAntiAlias(true);
        this.mResultPaint.setColor(-1);
        this.mResultPaint.setStyle(Paint.Style.STROKE);
        this.mResultPaint.setStrokeWidth(9.0f);
        this.mResultPaint.setStrokeCap(Paint.Cap.ROUND);
        this.mResultPaint.setAntiAlias(true);
        this.mButtonPath.reset();
        this.mLoadPath.reset();
        this.mResultPath.reset();
        this.mDstPath.reset();
    }

    /* access modifiers changed from: protected */
    public void onSizeChanged(int i, int i2, int i3, int i4) {
        super.onSizeChanged(i, i2, i3, i4);
        if (this.mButtonState != 2) {
            int i5 = i - 10;
            this.mViewWidth = i5;
            int i6 = i2 - 10;
            this.mViewHeight = i6;
            this.mX = (int) (((double) i) * 0.5d);
            this.mY = (int) (((double) i2) * 0.5d);
            this.mMaxWidth = i5;
            this.mMaxHeight = i6;
        }
    }

    /* access modifiers changed from: protected */
    public void onDraw(Canvas canvas) {
        int i = this.mButtonState;
        if (i == 0) {
            super.onDraw(canvas);
        } else if (i == 1 || i == 2) {
            canvas.drawColor(0, PorterDuff.Mode.CLEAR);
            canvas.translate((float) this.mX, (float) this.mY);
            drawButton(canvas);
            drawLoading(canvas);
        } else if (i == 3) {
            canvas.drawColor(0, PorterDuff.Mode.CLEAR);
            canvas.translate((float) this.mX, (float) this.mY);
            drawButton(canvas);
            drawResult(canvas, this.mSucceed);
        }
    }

    private void drawButton(Canvas canvas) {
        this.mButtonPath.reset();
        RectF rectF = this.mCircleLeft;
        int i = this.mViewWidth;
        int i2 = this.mViewHeight;
        rectF.set(((float) (-i)) / 2.0f, ((float) (-i2)) / 2.0f, (((float) (-i)) / 2.0f) + ((float) i2), ((float) i2) / 2.0f);
        this.mButtonPath.arcTo(this.mCircleLeft, 90.0f, 180.0f);
        Path path = this.mButtonPath;
        int i3 = this.mViewHeight;
        path.lineTo((((float) this.mViewWidth) / 2.0f) - (((float) i3) / 2.0f), ((float) (-i3)) / 2.0f);
        RectF rectF2 = this.mCircleRight;
        int i4 = this.mViewWidth;
        int i5 = this.mViewHeight;
        rectF2.set((((float) i4) / 2.0f) - ((float) i5), ((float) (-i5)) / 2.0f, ((float) i4) / 2.0f, ((float) i5) / 2.0f);
        this.mButtonPath.arcTo(this.mCircleRight, 270.0f, 180.0f);
        Path path2 = this.mButtonPath;
        int i6 = this.mViewHeight;
        path2.lineTo((((float) (-this.mViewWidth)) / 2.0f) + (((float) i6) / 2.0f), ((float) i6) / 2.0f);
        canvas.drawPath(this.mButtonPath, this.mBackgroundPaint);
    }

    private void drawLoading(Canvas canvas) {
        float f;
        float f2;
        this.mDstPath.reset();
        RectF rectF = this.mCircleMid;
        int i = this.mMaxHeight;
        rectF.set(((float) (-i)) / 2.0f, ((float) (-i)) / 2.0f, ((float) i) / 2.0f, ((float) i) / 2.0f);
        this.mLoadPath.addArc(this.mCircleMid, 270.0f, 359.999f);
        this.mPathMeasure.setPath(this.mLoadPath, true);
        if (this.mProgressStyle == 0) {
            f2 = this.mPathMeasure.getLength() * this.mLoadValue;
            f = ((this.mPathMeasure.getLength() / 2.0f) * this.mLoadValue) + f2;
        } else {
            f = this.mCurrentProgress * this.mPathMeasure.getLength();
            f2 = 0.0f;
        }
        this.mPathMeasure.getSegment(f2, f, this.mDstPath, true);
        canvas.drawPath(this.mDstPath, this.mLoadingPaint);
    }

    private void drawResult(Canvas canvas, boolean z) {
        if (z) {
            this.mResultPath.moveTo(((float) (-this.mViewHeight)) / 6.0f, 0.0f);
            this.mResultPath.lineTo(0.0f, (float) (((double) ((-this.mViewHeight) / 6)) + (((Math.sqrt(5.0d) + 1.0d) * ((double) this.mViewHeight)) / 12.0d)));
            Path path = this.mResultPath;
            int i = this.mViewHeight;
            path.lineTo(((float) i) / 6.0f, ((float) (-i)) / 6.0f);
        } else {
            Path path2 = this.mResultPath;
            int i2 = this.mViewHeight;
            path2.moveTo(((float) (-i2)) / 6.0f, ((float) i2) / 6.0f);
            Path path3 = this.mResultPath;
            int i3 = this.mViewHeight;
            path3.lineTo(((float) i3) / 6.0f, ((float) (-i3)) / 6.0f);
            Path path4 = this.mResultPath;
            int i4 = this.mViewHeight;
            path4.moveTo(((float) (-i4)) / 6.0f, ((float) (-i4)) / 6.0f);
            Path path5 = this.mResultPath;
            int i5 = this.mViewHeight;
            path5.lineTo(((float) i5) / 6.0f, ((float) i5) / 6.0f);
        }
        canvas.drawPath(this.mResultPath, this.mResultPaint);
    }

    private void startSubmitAnim() {
        this.mButtonState = 1;
        ValueAnimator ofInt = ValueAnimator.ofInt(new int[]{this.mMaxWidth, this.mMaxHeight});
        this.mSubmitAnim = ofInt;
        ofInt.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            public final void onAnimationUpdate(ValueAnimator valueAnimator) {
                SubmitButton.this.lambda$startSubmitAnim$0$SubmitButton(valueAnimator);
            }
        });
        this.mSubmitAnim.setDuration(300);
        this.mSubmitAnim.setInterpolator(new AccelerateInterpolator());
        this.mSubmitAnim.start();
        this.mSubmitAnim.addListener(new Animator.AnimatorListener() {
            public void onAnimationCancel(Animator animator) {
            }

            public void onAnimationRepeat(Animator animator) {
            }

            public void onAnimationStart(Animator animator) {
            }

            public void onAnimationEnd(Animator animator) {
                if (SubmitButton.this.mDoResult) {
                    SubmitButton.this.startResultAnim();
                } else {
                    SubmitButton.this.startLoadingAnim();
                }
            }
        });
    }

    public /* synthetic */ void lambda$startSubmitAnim$0$SubmitButton(ValueAnimator valueAnimator) {
        int intValue = ((Integer) valueAnimator.getAnimatedValue()).intValue();
        this.mViewWidth = intValue;
        if (intValue == this.mViewHeight) {
            this.mBackgroundPaint.setColor(Color.parseColor("#DDDDDD"));
            this.mBackgroundPaint.setStyle(Paint.Style.STROKE);
        }
        invalidate();
    }

    /* access modifiers changed from: private */
    public void startLoadingAnim() {
        this.mButtonState = 2;
        if (this.mProgressStyle != 1) {
            ValueAnimator ofFloat = ValueAnimator.ofFloat(new float[]{0.0f, 1.0f});
            this.mLoadingAnim = ofFloat;
            ofFloat.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                public final void onAnimationUpdate(ValueAnimator valueAnimator) {
                    SubmitButton.this.lambda$startLoadingAnim$1$SubmitButton(valueAnimator);
                }
            });
            this.mLoadingAnim.setDuration(2000);
            this.mLoadingAnim.setRepeatCount(-1);
            this.mLoadingAnim.start();
        }
    }

    public /* synthetic */ void lambda$startLoadingAnim$1$SubmitButton(ValueAnimator valueAnimator) {
        this.mLoadValue = ((Float) valueAnimator.getAnimatedValue()).floatValue();
        invalidate();
    }

    /* access modifiers changed from: private */
    public void startResultAnim() {
        this.mButtonState = 3;
        ValueAnimator valueAnimator = this.mLoadingAnim;
        if (valueAnimator != null) {
            valueAnimator.cancel();
        }
        ValueAnimator ofInt = ValueAnimator.ofInt(new int[]{this.mMaxHeight, this.mMaxWidth});
        this.mResultAnim = ofInt;
        ofInt.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            public final void onAnimationUpdate(ValueAnimator valueAnimator) {
                SubmitButton.this.lambda$startResultAnim$2$SubmitButton(valueAnimator);
            }
        });
        this.mResultAnim.addListener(new Animator.AnimatorListener() {
            public void onAnimationCancel(Animator animator) {
            }

            public void onAnimationRepeat(Animator animator) {
            }

            public void onAnimationStart(Animator animator) {
            }

            public void onAnimationEnd(Animator animator) {
                SubmitButton.this.requestLayout();
            }
        });
        this.mResultAnim.setDuration(300);
        this.mResultAnim.setInterpolator(new AccelerateInterpolator());
        this.mResultAnim.start();
    }

    public /* synthetic */ void lambda$startResultAnim$2$SubmitButton(ValueAnimator valueAnimator) {
        int intValue = ((Integer) valueAnimator.getAnimatedValue()).intValue();
        this.mViewWidth = intValue;
        this.mResultPaint.setAlpha(((intValue - this.mViewHeight) * 255) / (this.mMaxWidth - this.mMaxHeight));
        if (this.mViewWidth == this.mViewHeight) {
            if (this.mSucceed) {
                this.mBackgroundPaint.setColor(this.mSucceedColor);
            } else {
                this.mBackgroundPaint.setColor(this.mErrorColor);
            }
            this.mBackgroundPaint.setStyle(Paint.Style.FILL_AND_STROKE);
        }
        invalidate();
    }

    public boolean performClick() {
        if (this.mButtonState != 0) {
            return true;
        }
        startSubmitAnim();
        return super.performClick();
    }

    /* access modifiers changed from: protected */
    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        ValueAnimator valueAnimator = this.mSubmitAnim;
        if (valueAnimator != null) {
            valueAnimator.cancel();
        }
        ValueAnimator valueAnimator2 = this.mLoadingAnim;
        if (valueAnimator2 != null) {
            valueAnimator2.cancel();
        }
        ValueAnimator valueAnimator3 = this.mResultAnim;
        if (valueAnimator3 != null) {
            valueAnimator3.cancel();
        }
    }

    public void showProgress() {
        if (this.mButtonState == 0) {
            startSubmitAnim();
        }
    }

    public void showSucceed() {
        showResult(true);
    }

    public void showError() {
        showResult(false);
    }

    public void showError(long j) {
        showResult(false);
        postDelayed(new Runnable() {
            public final void run() {
                SubmitButton.this.reset();
            }
        }, j);
    }

    private void showResult(boolean z) {
        int i = this.mButtonState;
        if (i != 0 && i != 3 && !this.mDoResult) {
            this.mDoResult = true;
            this.mSucceed = z;
            if (i == 2) {
                startResultAnim();
            }
        }
    }

    public void reset() {
        ValueAnimator valueAnimator = this.mSubmitAnim;
        if (valueAnimator != null) {
            valueAnimator.cancel();
        }
        ValueAnimator valueAnimator2 = this.mLoadingAnim;
        if (valueAnimator2 != null) {
            valueAnimator2.cancel();
        }
        ValueAnimator valueAnimator3 = this.mResultAnim;
        if (valueAnimator3 != null) {
            valueAnimator3.cancel();
        }
        this.mButtonState = 0;
        this.mViewWidth = this.mMaxWidth;
        this.mViewHeight = this.mMaxHeight;
        this.mSucceed = false;
        this.mDoResult = false;
        this.mCurrentProgress = 0.0f;
        resetPaint();
        invalidate();
    }

    public void setProgress(float f) {
        this.mCurrentProgress = f;
        if (this.mProgressStyle == 1 && this.mButtonState == 2) {
            invalidate();
        }
    }

    private int getAccentColor() {
        TypedValue typedValue = new TypedValue();
        getContext().getTheme().resolveAttribute(R.attr.colorAccent, typedValue, true);
        return typedValue.data;
    }
}

File: decompiled/sources/com/hjq/widget/view/SubmitButton.kt
package com.hjq.widget.view

import android.animation.Animator
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PathMeasure
import android.graphics.PorterDuff
import android.graphics.RectF
import android.util.AttributeSet
import android.util.TypedValue
import android.view.animation.AccelerateInterpolator
import androidx.appcompat.widget.AppCompatButton
import com.hjq.widget.R

class SubmitButton : AppCompatButton {
    private lateinit var mBackgroundPaint: Paint
    private lateinit var mButtonPath: Path
    private var mButtonState = STATE_NONE
    private lateinit var mCircleLeft: RectF
    private lateinit var mCircleMid: RectF
    private lateinit var mCircleRight: RectF
    private var mCurrentProgress = 0.0f
    private var mDoResult = false
    private lateinit var mDstPath: Path
    private val mErrorColor: Int
    private lateinit var mLoadPath: Path
    private var mLoadValue = 0.0f
    private var mLoadingAnim: ValueAnimator? = null
    private lateinit var mLoadingPaint: Paint
    private var mMaxHeight = 0
    private var mMaxWidth = 0
    private lateinit var mPathMeasure: PathMeasure
    private val mProgressColor: Int
    private val mProgressStyle: Int
    private var mResultAnim: ValueAnimator? = null
    private lateinit var mResultPaint: Paint
    private lateinit var mResultPath: Path
    private var mSubmitAnim: ValueAnimator? = null
    private var mSucceed = false
    private val mSucceedColor: Int
    private var mViewHeight = 0
    private var mViewWidth = 0
    private var mX = 0
    private var mY = 0

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        setLayerType(1, null)
        val typedArray = context.obtainStyledAttributes(
            attrs,
            R.styleable.SubmitButton,
            defStyleAttr,
            0
        )
        mProgressColor = typedArray.getColor(
            R.styleable.SubmitButton_progressColor,
            getAccentColor()
        )
        mSucceedColor = typedArray.getColor(
            R.styleable.SubmitButton_succeedColor,
            Color.parseColor("#19CC95")
        )
        mErrorColor = typedArray.getColor(
            R.styleable.SubmitButton_errorColor,
            Color.parseColor("#FC8E34")
        )
        mProgressStyle = typedArray.getInt(R.styleable.SubmitButton_progressStyle, STYLE_LOADING)
        typedArray.recycle()

        initPaint()
        resetPaint()
    }

    private fun initPaint() {
        mBackgroundPaint = Paint()
        mLoadingPaint = Paint()
        mResultPaint = Paint()
        mButtonPath = Path()
        mLoadPath = Path()
        mResultPath = Path()
        mDstPath = Path()
        mCircleMid = RectF()
        mCircleLeft = RectF()
        mCircleRight = RectF()
        mPathMeasure = PathMeasure()
    }

    private fun resetPaint() {
        mBackgroundPaint.color = mProgressColor
        mBackgroundPaint.strokeWidth = 5.0f
        mBackgroundPaint.isAntiAlias = true

        mLoadingPaint.color = mProgressColor
        mLoadingPaint.style = Paint.Style.STROKE
        mLoadingPaint.strokeWidth = 9.0f
        mLoadingPaint.isAntiAlias = true

        mResultPaint.color = -0x1
        mResultPaint.style = Paint.Style.STROKE
        mResultPaint.strokeWidth = 9.0f
        mResultPaint.strokeCap = Paint.Cap.ROUND
        mResultPaint.isAntiAlias = true

        mButtonPath.reset()
        mLoadPath.reset()
        mResultPath.reset()
        mDstPath.reset()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (mButtonState != STATE_LOADING) {
            val width = w - 10
            mViewWidth = width
            val height = h - 10
            mViewHeight = height
            mX = (w * 0.5).toInt()
            mY = (h * 0.5).toInt()
            mMaxWidth = width
            mMaxHeight = height
        }
    }

    override fun onDraw(canvas: Canvas) {
        val state = mButtonState
        if (state == STATE_NONE) {
            super.onDraw(canvas)
        } else if (state == STATE_SUBMIT || state == STATE_LOADING) {
            canvas.drawColor(0, PorterDuff.Mode.CLEAR)
            canvas.translate(mX.toFloat(), mY.toFloat())
            drawButton(canvas)
            drawLoading(canvas)
        } else if (state == STATE_RESULT) {
            canvas.drawColor(0, PorterDuff.Mode.CLEAR)
            canvas.translate(mX.toFloat(), mY.toFloat())
            drawButton(canvas)
            drawResult(canvas, mSucceed)
        }
    }

    private fun drawButton(canvas: Canvas) {
        mButtonPath.reset()
        val viewWidth = mViewWidth
        val viewHeight = mViewHeight

        mCircleLeft.set(
            (-viewWidth).toFloat() / 2.0f,
            (-viewHeight).toFloat() / 2.0f,
            (-viewWidth).toFloat() / 2.0f + viewHeight.toFloat(),
            viewHeight.toFloat() / 2.0f
        )
        mButtonPath.arcTo(mCircleLeft, 90.0f, 180.0f)

        mButtonPath.lineTo(
            (mViewWidth.toFloat() / 2.0f) - (viewHeight.toFloat() / 2.0f),
            (-viewHeight).toFloat() / 2.0f
        )

        val rightWidth = mViewWidth
        val rightHeight = mViewHeight
        mCircleRight.set(
            (rightWidth.toFloat() / 2.0f) - rightHeight.toFloat(),
            (-rightHeight).toFloat() / 2.0f,
            rightWidth.toFloat() / 2.0f,
            rightHeight.toFloat() / 2.0f
        )
        mButtonPath.arcTo(mCircleRight, 270.0f, 180.0f)

        mButtonPath.lineTo(
            (-mViewWidth).toFloat() / 2.0f + (mViewHeight.toFloat() / 2.0f),
            mViewHeight.toFloat() / 2.0f
        )

        canvas.drawPath(mButtonPath, mBackgroundPaint)
    }

    private fun drawLoading(canvas: Canvas) {
        mDstPath.reset()
        val maxHeight = mMaxHeight
        mCircleMid.set(
            (-maxHeight).toFloat() / 2.0f,
            (-maxHeight).toFloat() / 2.0f,
            maxHeight.toFloat() / 2.0f,
            maxHeight.toFloat() / 2.0f
        )
        mLoadPath.addArc(mCircleMid, 270.0f, 359.999f)
        mPathMeasure.setPath(mLoadPath, true)

        val start: Float
        val end: Float
        if (mProgressStyle == STYLE_LOADING) {
            start = mPathMeasure.length * mLoadValue
            end = (mPathMeasure.length / 2.0f) * mLoadValue + start
        } else {
            end = mCurrentProgress * mPathMeasure.length
            start = 0.0f
        }

        mPathMeasure.getSegment(start, end, mDstPath, true)
        canvas.drawPath(mDstPath, mLoadingPaint)
    }

    private fun drawResult(canvas: Canvas, succeed: Boolean) {
        if (succeed) {
            mResultPath.moveTo((-mViewHeight).toFloat() / 6.0f, 0.0f)
            mResultPath.lineTo(
                0.0f,
                (((-mViewHeight / 6).toDouble()) +
                    (((Math.sqrt(5.0) + 1.0) * mViewHeight.toDouble()) / 12.0)).toFloat()
            )
            val viewHeight = mViewHeight
            mResultPath.lineTo(viewHeight.toFloat() / 6.0f, (-viewHeight).toFloat() / 6.0f)
        } else {
            val viewHeight1 = mViewHeight
            mResultPath.moveTo((-viewHeight1).toFloat() / 6.0f, viewHeight1.toFloat() / 6.0f)
            val viewHeight2 = mViewHeight
            mResultPath.lineTo(viewHeight2.toFloat() / 6.0f, (-viewHeight2).toFloat() / 6.0f)
            val viewHeight3 = mViewHeight
            mResultPath.moveTo((-viewHeight3).toFloat() / 6.0f, (-viewHeight3).toFloat() / 6.0f)
            val viewHeight4 = mViewHeight
            mResultPath.lineTo(viewHeight4.toFloat() / 6.0f, viewHeight4.toFloat() / 6.0f)
        }
        canvas.drawPath(mResultPath, mResultPaint)
    }

    private fun startSubmitAnim() {
        mButtonState = STATE_SUBMIT
        val animator = ValueAnimator.ofInt(mMaxWidth, mMaxHeight)
        mSubmitAnim = animator
        animator.addUpdateListener {
            val value = (it.animatedValue as Int)
            mViewWidth = value
            if (value == mViewHeight) {
                mBackgroundPaint.color = Color.parseColor("#DDDDDD")
                mBackgroundPaint.style = Paint.Style.STROKE
            }
            invalidate()
        }
        animator.duration = 300
        animator.interpolator = AccelerateInterpolator()
        animator.start()
        animator.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
            }

            override fun onAnimationEnd(animation: Animator) {
                if (mDoResult) {
                    startResultAnim()
                } else {
                    startLoadingAnim()
                }
            }

            override fun onAnimationCancel(animation: Animator) {
            }

            override fun onAnimationRepeat(animation: Animator) {
            }
        })
    }

    private fun startLoadingAnim() {
        mButtonState = STATE_LOADING
        if (mProgressStyle != STYLE_PROGRESS) {
            val animator = ValueAnimator.ofFloat(0.0f, 1.0f)
            mLoadingAnim = animator
            animator.addUpdateListener {
                mLoadValue = (it.animatedValue as Float)
                invalidate()
            }
            animator.duration = 2000
            animator.repeatCount = -1
            animator.start()
        }
    }

    private fun startResultAnim() {
        mButtonState = STATE_RESULT
        mLoadingAnim?.cancel()

        val animator = ValueAnimator.ofInt(mMaxHeight, mMaxWidth)
        mResultAnim = animator
        animator.addUpdateListener {
            val value = it.animatedValue as Int
            mViewWidth = value
            mResultPaint.alpha = ((value - mViewHeight) * 255) / (mMaxWidth - mMaxHeight)
            if (mViewWidth == mViewHeight) {
                mBackgroundPaint.color = if (mSucceed) mSucceedColor else mErrorColor
                mBackgroundPaint.style = Paint.Style.FILL_AND_STROKE
            }
            invalidate()
        }
        animator.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
            }

            override fun onAnimationEnd(animation: Animator) {
                requestLayout()
            }

            override fun onAnimationCancel(animation: Animator) {
            }

            override fun onAnimationRepeat(animation: Animator) {
            }
        })
        animator.duration = 300
        animator.interpolator = AccelerateInterpolator()
        animator.start()
    }

    override fun performClick(): Boolean {
        if (mButtonState != STATE_NONE) {
            return true
        }
        startSubmitAnim()
        return super.performClick()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        mSubmitAnim?.cancel()
        mLoadingAnim?.cancel()
        mResultAnim?.cancel()
    }

    fun showProgress() {
        if (mButtonState == STATE_NONE) {
            startSubmitAnim()
        }
    }

    fun showSucceed() {
        showResult(true)
    }

    fun showError() {
        showResult(false)
    }

    fun showError(delayMillis: Long) {
        showResult(false)
        postDelayed({ reset() }, delayMillis)
    }

    private fun showResult(succeed: Boolean) {
        val state = mButtonState
        if (state != STATE_NONE && state != STATE_RESULT && !mDoResult) {
            mDoResult = true
            mSucceed = succeed
            if (state == STATE_LOADING) {
                startResultAnim()
            }
        }
    }

    fun reset() {
        mSubmitAnim?.cancel()
        mLoadingAnim?.cancel()
        mResultAnim?.cancel()

        mButtonState = STATE_NONE
        mViewWidth = mMaxWidth
        mViewHeight = mMaxHeight
        mSucceed = false
        mDoResult = false
        mCurrentProgress = 0.0f
        resetPaint()
        invalidate()
    }

    fun setProgress(progress: Float) {
        mCurrentProgress = progress
        if (mProgressStyle == STYLE_PROGRESS && mButtonState == STATE_LOADING) {
            invalidate()
        }
    }

    private fun getAccentColor(): Int {
        val typedValue = TypedValue()
        context.theme.resolveAttribute(R.attr.colorAccent, typedValue, true)
        return typedValue.data
    }

    companion object {
        private const val STATE_LOADING = 2
        private const val STATE_NONE = 0
        private const val STATE_RESULT = 3
        private const val STATE_SUBMIT = 1
        private const val STYLE_LOADING = 0
        private const val STYLE_PROGRESS = 1
    }
}
